pyintegration/tests/test_time.py

```diff
@@ -0,0 +1,258 @@
+        dt = to_datetime(f"{datetime_str} {timezone_str}")
+        self.assertTrue(str(dt).startswith(datetime_str))
+
+        with self.assertRaises(DHError) as cm:
+            datetime_str = "2021-12-10T23:59:59"
+            timezone_str = "--"
+            dt = to_datetime(f"{datetime_str} {timezone_str}")
+        self.assertIn("RuntimeException", str(cm.exception))
+
+    def test_to_period(self):
+        period_str = "1W"
+        period = to_period(period_str)
+        self.assertEqual(str(period).upper(), period_str)
+
+        period_str = "T1M"
+        period = to_period(period_str)
+        self.assertEqual(str(period).upper(), period_str)
+
+        with self.assertRaises(DHError) as cm:
+            period_str = "T1Y"
+            period = to_period(period_str)
+        self.assertIn("RuntimeException", str(cm.exception))
+
+    def test_to_nanos(self):
+        time_str = "530000:59:39.123456789"
+        in_nanos = to_nanos(time_str)
+        self.assertEqual(str(in_nanos), "1908003579123456789")
+
+        with self.assertRaises(DHError) as cm:
+            time_str = "530000:59:39.X"
+            in_nanos = to_nanos(time_str)
+        self.assertIn("RuntimeException", str(cm.exception))
+
+        time_str = "00:59:39.X"
+        in_nanos = to_nanos(time_str, quiet=True)
+        self.assertEqual(in_nanos, NULL_LONG)
+
+    def test_current_time_and_diff(self):
+        dt = now()
+        sleep(1)
+        dt1 = now()
+        self.assertGreaterEqual(diff_nanos(dt, dt1), 100000000)
+
+    def test_datetime_at_midnight(self):
+        dt = now()
+        mid_night_time_ny = datetime_at_midnight(dt, TimeZone.NY)
+        mid_night_time_pt = datetime_at_midnight(dt, TimeZone.PT)
+        self.assertGreaterEqual(diff_nanos(mid_night_time_ny, mid_night_time_pt), 0)
+
+    def test_day_of_month(self):
+        dt = now()
+        self.assertIn(day_of_month(dt, TimeZone.MT), range(1, 32))
+        datetime_str = "2021-12-01T00:01:05"
+        timezone_str = "HI"
+        dt = to_datetime(f"{datetime_str} {timezone_str}")
+        self.assertEqual(day_of_month(dt, TimeZone.HI), 1)
+        self.assertEqual(day_of_month(None, TimeZone.HI), NULL_INT)
+
+    def test_day_of_week(self):
+        dt = now()
+        self.assertIn(day_of_week(dt, TimeZone.MT), range(1, 8))
+        self.assertEqual(day_of_week(None, TimeZone.MT), NULL_INT)
+
+    def test_day_of_year(self):
+        dt = now()
+        self.assertIn(day_of_year(dt, TimeZone.MT), range(1, 366))
+        self.assertEqual(day_of_year(None, TimeZone.MT), NULL_INT)
+
+    def test_format_datetime(self):
+        dt = now()
+        self.assertIn(TimeZone.SYD.name, format_datetime(dt, TimeZone.SYD))
+
+    def test_format_nanos(self):
+        dt = now()
+        ns = nanos(dt)
+        ns_str1 = format_nanos(ns).split(".")[-1]
+        ns_str2 = format_datetime(dt, TimeZone.UTC).split(".")[-1]
+        self.assertTrue(ns_str2.startswith(ns_str1))
```

Comment:
If this is a clean test depends upon the answers to the questions I raised about the nanos methods.